File: handy-js/quiz/q20_linked_list_cycle_easy.js
/**
 * Q20) Linked List Cycle
 *
 * AI-BEST:
 * Floyd's tortoise-hare pointers.
 * Time: O(n), Space: O(1)
 */

class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function hasCycleBest(head) {
  // Step 1: Use two pointers with different speeds.
  // Why: if a cycle exists, fast eventually laps slow and they meet.
  let slow = head;
  let fast = head;

  // Step 2: Move until fast hits end (no cycle) or meets slow (cycle).
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
    if (slow === fast) return true;
  }

  // Step 3: Reaching null means no cycle.
  return false;
}

function buildLinkedList(values, cyclePos) {
  if (!values.length) return null;

  const nodes = values.map((v) => new ListNode(v));
  for (let i = 0; i < nodes.length - 1; i++) {
    nodes[i].next = nodes[i + 1];
  }

  if (cyclePos >= 0 && cyclePos < nodes.length) {
    nodes[nodes.length - 1].next = nodes[cyclePos];
  }

  return nodes[0];
}

function runDemo() {
  console.log("Q20: Linked List Cycle");
  const cases = [
    { values: [3, 2, 0, -4], cyclePos: 1 },
    { values: [1, 2], cyclePos: 0 },
    { values: [1], cyclePos: -1 },
  ];

  for (const { values, cyclePos } of cases) {
    const head = buildLinkedList(values, cyclePos);
    console.log(`Input: values=${JSON.stringify(values)}, cyclePos=${cyclePos}`);
    console.log("  BEST:", hasCycleBest(head));
  }
}

runDemo();

File: handy-java/quiz/q20_linked_list_cycle_easy.java
import java.util.Arrays;

/**
 * Q20) Linked List Cycle
 *
 * AI-BEST:
 * Floyd's slow/fast pointers.
 * Time: O(n), Space: O(1)
 */
public class q20_linked_list_cycle_easy {
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
        }
    }

    static boolean hasCycleBest(ListNode head) {
        // Step 1: Move slow by 1 and fast by 2.
        // Why: in a cycle, fast eventually catches slow.
        ListNode slow = head;
        ListNode fast = head;

        // Step 2: Stop on meet (cycle) or null (no cycle).
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true;
        }

        // Step 3: Reaching list end means no cycle.
        return false;
    }

    static ListNode buildListWithCycle(int[] values, int cyclePos) {
        if (values.length == 0) return null;
        ListNode[] nodes = new ListNode[values.length];
        for (int i = 0; i < values.length; i++) nodes[i] = new ListNode(values[i]);
        for (int i = 0; i < values.length - 1; i++) nodes[i].next = nodes[i + 1];
        if (cyclePos >= 0 && cyclePos < values.length) {
            nodes[values.length - 1].next = nodes[cyclePos];
        }
        return nodes[0];
    }

    public static void main(String[] args) {
        System.out.println("Q20: Linked List Cycle");
        int[][] valueCases = {
                {3, 2, 0, -4},
                {1, 2},
                {1}
        };
        int[] cyclePosCases = {1, 0, -1};

        for (int i = 0; i < valueCases.length; i++) {
            int[] values = valueCases[i];
            int cyclePos = cyclePosCases[i];
            ListNode head = buildListWithCycle(values, cyclePos);
            System.out.println("Input: values=" + Arrays.toString(values) + ", cyclePos=" + cyclePos);
            System.out.println("  BEST: " + hasCycleBest(head));
        }
    }
}
